src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1084 @@
+                }
+
+                int index = patternIndex;
+                foreach ((BoundPattern pattern, SyntaxNode? syntax) in set.Cases)
+                {
+                    var label = new GeneratedLabelSymbol("orCase");
+                    SyntaxNode? diagSyntax = syntax;
+                    if (diagSyntax is null)
+                    {
+                        labelsToIgnore.Add(label);
+                        diagSyntax = nodeSyntax;
+                    }
+
+                    Debug.Assert(diagSyntax is not null);
+                    casesBuilder.Add(builder.MakeTestsForPattern(++index, diagSyntax, rootIdentifier, pattern, whenClause: null, label: label));
+                }
+            }
+        }
+
+        /// <summary>
+        /// The purpose of this method is to bring `or` sequences to the top-level (so they can be used as separate cases).
+        /// Each set handles one `or`.
+        /// </summary>
+        private static SetsOfOrCases RewriteToSetsOfOrCases(BoundPattern? pattern, Conversions conversions)
+        {
+            return pattern switch
+            {
+                BoundBinaryPattern binary => rewriteBinary(binary, conversions),
+                BoundRecursivePattern => default,
+                BoundListPattern => default,
+                BoundSlicePattern => default,
+                BoundITuplePattern => default,
+                BoundNegatedPattern => default,
+                BoundTypePattern => default,
+                BoundDeclarationPattern => default,
+                BoundConstantPattern => default,
+                BoundDiscardPattern => default,
+                BoundRelationalPattern => default,
+                null => default,
+                _ => throw ExceptionUtilities.UnexpectedValue(pattern)
+            };
+
+            static SetsOfOrCases rewriteBinary(BoundBinaryPattern binaryPattern, Conversions conversions)
+            {
+                SetsOfOrCases result = default;
+
+                if (binaryPattern.Disjunction)
+                {
+                    var patterns = ArrayBuilder<BoundPattern>.GetInstance();
+                    addPatternsFromOrTree(binaryPattern, patterns);
+
+                    // In `A1 or ... or An`, we produce an expansion set: `A1`, ..., `An`
+                    OrCases resultOrSet1 = result.StartNewOrCases();
+                    var inputType = binaryPattern.InputType;
+                    foreach (var pattern in patterns)
+                    {
+                        resultOrSet1.Add(pattern, pattern.Syntax);
+                    }
+
+                    // If any of the nested patterns can be expanded, we carry those on.
+                    // For example, in `... or Ai or ...`, when we found multiple `or` patterns in `Ai` to be expanded
+                    // For each such nested expansion, we'll produce an `... or <expansion> or ...` expansion
+                    for (int i = 0; i < patterns.Count; i++)
```

Comment:
>for (int i = 0; i < patterns.Count; i++) [](http://example.com/codeflow?start=20&length=40)

It feels like there is still some excessiveness in the sets that we are producing in this loop. And I think this is causing some strange behavior for this scenario:
```
        [Fact]
        public void Test_07()
        {
            var source = """
class C
{
    void M(S s)
    {
#line 100
        if (s is { Prop0: 0 } or (({ Prop1: > 1 } or { Prop2: 2 }) and { Prop3: 3 }) or { Prop2: > -1 }) { }

#line 200
        if (s is { Prop0: 0 } or (({ Prop2: 2 } or { Prop1: > 1 }) and { Prop3: 3 }) or { Prop2: > -1 }) { }

#line 300
        if (s is { Prop0: 0 } or (({ Prop1: > 1 } or { Prop2: 2 }) and { Prop3: 3 }) ) { }
    }
}

public struct S
{
    public int Prop0 { get; set; }
    public int Prop1 { get; set; }
    public int Prop2 { get; set; }
    public int Prop3 { get; set; }
}
""";
            var compilation = CreateCompilation(source);
            compilation.VerifyDiagnostics(
                // (100,63): warning CS9269: The pattern is redundant.
                //         if (s is { Prop0: 0 } or (({ Prop1: > 1 } or { Prop2: 2 }) and { Prop3: 3 }) or { Prop2: > -1 }) { }
                Diagnostic(ErrorCode.WRN_RedundantPattern, "2").WithLocation(100, 63)
                );
        }
```

Note, we are reporting redundancy of a preceding pattern due to a pattern that comes later. The warning is reported for the first `if`, but not reported for the 3rd `if`. The only difference between the two is that the tailing ```or { Prop2: > -1 }``` is dropped in the the 3rd `if`. 
Also, notice that flipping the order of the reported pattern with another unrelated pattern makes the warning go away. There is no warning for the 2nd `if`, which flips the order of patterns in ```{ Prop1: > 1 } or { Prop2: 2 })```.

While, perhaps, one might argue that that pattern is redundant, the detection strategy that we are trying to implement is reliable only for detecting redundancy in the patterns that come later due to the patterns that come before.  Therefore, at this point I would like to suggest to not report/detect "backward" redundancy.

Here are the sets that we produce for ```{ Prop0: 0 } or (({ Prop1: > 1 } or { Prop2: 2 }) and { Prop3: 3 }) or { Prop2: > -1 }```:
```
Set:
{Prop0: 0}  => 0, 
(({Prop1: > 1} or {Prop2: 2}) and {Prop3: 3})  => ({ Prop1: > 1 } or { Prop2: 2 }) and { Prop3: 3 }, 
{Prop2: > -1}  => > -1, 

Set:
({Prop0: 0} or ({Prop1: > 1} or {Prop2: > -1}))  => > 1, 
({Prop0: 0} or ({Prop2: 2} or {Prop2: > -1}))  => 2, 
```

According to the proposed strategy, I would expect us to have the following states:
``` 
Set:
{Prop0: 0}  => 0, 
(({Prop1: > 1} or {Prop2: 2}) and {Prop3: 3})  => ({ Prop1: > 1 } or { Prop2: 2 }) and { Prop3: 3 }, 
{Prop2: > -1}  => > -1, 

Set:
({Prop0: 0} or {Prop1: > 1})  => > 1, 
({Prop0: 0} or {Prop2: 2})  => 2, 
```

or
``` 
Set:
{Prop0: 0}  => 0, 
(({Prop1: > 1} or {Prop2: 2}) and {Prop3: 3})  => ({ Prop1: > 1 } or { Prop2: 2 }) and { Prop3: 3 }, 
{Prop2: > -1}  => > -1, 

Set:
{Prop0: 0}  => 0, 
{Prop1: > 1}  => > 1, 
{Prop2: 2}  => 2, 
```

Reply:
Here is another example where we are failing to detect "forward" redundancy:
```
        [Fact]
        public void Test_08()
        {
            var source = """
class C
{
    void M(S s)
    {
#line 400
        if (s is { Prop0: 0 } or { Prop2: > -1 } or (({ Prop1: > 1 } or { Prop2: 2 }) and { Prop3: 3 })) { }

#line 500
        if (s is { Prop0: 0 } or { Prop2: > -1 } or (({ Prop2: 2 } or { Prop1: > 1 }) and { Prop3: 3 })) { }
    }
}

public struct S
{
    public int Prop0 { get; set; }
    public int Prop1 { get; set; }
    public int Prop2 { get; set; }
    public int Prop3 { get; set; }
}
""";
            var compilation = CreateCompilation(source);
            compilation.VerifyDiagnostics(
                // (400,82): warning CS9269: The pattern is redundant.
                //         if (s is { Prop0: 0 } or { Prop2: > -1 } or (({ Prop1: > 1 } or { Prop2: 2 }) and { Prop3: 3 })) { }
                Diagnostic(ErrorCode.WRN_RedundantPattern, "2").WithLocation(400, 82)
                );
        }
```

Redundancy of ```{ Prop2: 2 }``` is reported only for the first ```if```. I think the same excessiveness in the sets is to blame.
For ```{ Prop0: 0 } or { Prop2: > -1 } or (({ Prop2: 2 } or { Prop1: > 1 }) and { Prop3: 3 })``` we have the following sets:
```
Set:
{Prop0: 0}  => 0, 
{Prop2: > -1}  => > -1, 
(({Prop2: 2} or {Prop1: > 1}) and {Prop3: 3})  => ({ Prop2: 2 } or { Prop1: > 1 }) and { Prop3: 3 }, 

Set:
({Prop0: 0} or ({Prop2: > -1} or {Prop2: 2}))  => 2, 
({Prop0: 0} or ({Prop2: > -1} or {Prop1: > 1}))  => > 1, 
```

I would expect us to have the following sets:
```
Set:
{Prop0: 0}  => 0, 
{Prop2: > -1}  => > -1, 
(({Prop2: 2} or {Prop1: > 1}) and {Prop3: 3})  => ({ Prop2: 2 } or { Prop1: > 1 }) and { Prop3: 3 }, 

Set:
{Prop0: 0}  => 0, 
{Prop2: > -1}  => > -1, 
{Prop2: 2}  => 2, 
{Prop1: > 1}  => > 1, 
```

Reply:
Here is the implementation that I have in mind:
```
for (int i = 0; i < patterns.Count; i++)
{
    BoundPattern pattern = patterns[i];
    using SetsOfOrCases setOfOrCases = RewriteToSetsOfOrCases(pattern, conversions);
    if (!setOfOrCases.IsDefault)
    {
        foreach (OrCases orSet in setOfOrCases.Set)
        {
            OrCases resultOrSet = result.StartNewOrCases();

            for (int j = 0; j < i; j++)
            {
                resultOrSet.Add(patterns[j], syntax: null); // already checked reachability in the previous set
            }

            foreach ((BoundPattern resultPattern, SyntaxNode? syntax) in orSet.Cases)
            {
                resultOrSet.Add(resultPattern, syntax);
            }
        }
    }
}
```

Reply:
Thanks for the detailed comment. I stared at it a bunch, worked through expansion manually and finally understood. 

Tried a few more expansion examples, such as:
```
        if (s is { P0: 0 }
            or { P2: > -1 }
            or (({ P1: > 1 } or { P2: > -1 }) and { P3: 3 })
            or (({ P4: > 1 } or { P2: 2 }) and { P3: 3 })) { }

Set:
{P0: 0}  => 0, 
{P2: > -1}  => > -1, 
(({P1: > 1} or {P2: > -1}) and {P3: 3})  => ({ P1: > 1 } or { P2: > -1 }) and { P3: 3 }, 
(({P4: > 1} or {P2: 2}) and {P3: 3})  => ({ P4: > 1 } or { P2: 2 }) and { P3: 3 }, 

Set:
{P0: 0}
{P2: > -1}
{P1: > 1}  => > 1, 
{P2: > -1}  => > -1, 

Set:
{P0: 0}
{P2: > -1}
(({P1: > 1} or {P2: > -1}) and {P3: 3})
{P4: > 1}  => > 1, 
{P2: 2}  => 2, 
```